### PR TITLE
feat(ingest): Add Metadata Ingestion Source Clickhouse support

### PR DESCRIPTION
**Pull request by havramar**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+
+                    try:
+                        client_object_type = service.get_column_by_spec(
+                            column_raw,
+                            {
+                                'context': self._client.connection.context
+                            }
+                        )
+                    except UnknownTypeError as e:
+                        if 'bool' == column_raw.lower():
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:188]: Is this mapping a boolean type to an int8 one?

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    String.__name__: StringTypeClass,
+    ByteString.__name__: StringTypeClass,
+    FixedString.__name__: StringTypeClass,
+    ByteFixedString.__name__: StringTypeClass,
+
+    TupleColumn.__name__: ArrayTypeClass,
+
+    UUIDColumn.__name__: StringTypeClass,
+
+    'Bool': NumberTypeClass
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:300]: We do have a `BooleanTypeClass` that may be more suitable here

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+    IPv4Column.__name__: NumberTypeClass,
+    IPv6Column.__name__: NumberTypeClass,
+
+    LowCardinalityColumn.__name__: RecordTypeClass,
+
+    MapColumn.__name__: MapTypeClass,
+
+    String.__name__: StringTypeClass,
+    ByteString.__name__: StringTypeClass,
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:292]: There's a `BytesTypeClass` that could fit here—though I'm not sure what a ByteString is

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        platform = "clickhouse"
+        databases = self._client.execute('show databases')
+        for database, in databases:
+            if not self.config.schema_pattern.allowed(database):
+                logger.info(f'Skipping database: {database}')
+                continue
+
+            tables = self._client.execute(f'show tables from {database}')
+            for table, in tables:
+                print(table)
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:156]: Nit: maybe convert this to a `.info()` and add some description so we know it's a table being printed.

File: metadata-ingestion/source_docs/clickhouse.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+## Setup
+
+To install this plugin, run `pip install 'acryl-datahub[clickhouse]'`.
+
+## Capabilities
+
+This plugin extracts the following:
+
+- Metadata for tables, views and materialized views
```
Comment (kevinhu) [inline metadata-ingestion/source_docs/clickhouse.md:13]: I'm not sure if I saw any queries for views/materialized views in the logic. Are these included by default when calling `show tables from {database}` ?

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+                # 1. Prepare DatasetSnapshot
+                dataset_snapshot = DatasetSnapshot(
+                    urn=make_dataset_urn(platform,
+                                         dataset_name,
+                                         self.config.env),
+                    aspects=[],
+                )
+
+                # 3. Prepare SchemaFields for SchemaMetadata
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:171]: Nit: Missing a step 2

File: metadata-ingestion/src/datahub/ingestion/source/clickhouse.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                            client_object_type = Int8Column()
+                        elif column_raw.startswith('AggregateFunction'):
+                            client_object_type = IntColumn()
+                        else:
+                            client_object_type = String()
+
+                    try:
+                        datahub_type = _field_type_mapping_clickhouse[client_object_type.__class__.__name__]
+                    except KeyError:
+                        logger.info(f'Could not map column type: {database}.{table}.{column_raw}')
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/clickhouse.py:198]: May want to use `logger.warning` here

**Review by kevinhu — COMMENTED** [review]
Looks good! Just a few comments on the details. Thanks for writing this!

**Comment by havramar** [comment]
@treff7es thank you for your feedback!

Today better approach (based on sqlalchemy) showed up in new PR: https://github.com/linkedin/datahub/pull/4057
So there is no need to keep mine, so I'm closing it.

**Pull request closed without merging**